### PR TITLE
[rocm] Excises (almost) dependence on /opt/rocm from the compiler.

### PR DESCRIPTION
**Pull request by stellaraccident**
Almost all uses were for bitcode linking, which we have been bundling with the compiler for a long time.

I left one ancillary use in the ukernel directory as that depends on some HIP headers. We can sever that dependency too, but will do that in a separate PR as it isn't related.

Also takes this chance to:

* Remove foot-guns around the use of `--iree-rocm-link-bc` flag by defaulting it to true. Afaict, it was only false because in the early phases it was hard to ensure that everyone had bitcode files available.
* Remove all CMake plumbing that was setting `--iree-rocm-link-bc` or `--iree-rocm-bc-dir`, instead making the build system work with defaults for tests.
* Revamped error handling logic in ROCMTarget, which was really quite awful. It was spewing errors to stderr and just barreling on in some situations. All failure messages have been moved to the point encountered and handled properly with an MLIR Location.
* The above uncovered a bug where we were segfaulting on tests that were missing an ordinal attribute. These now properly error, and I fixed the tests.
* Deletes the ROCM bazel lit tests since Bazel lacks the infrastructure to bundle bitcode files and will likely never get it. We prefer to have the build exercise default flag flows which precludes these tests being the same between these systems.

I can't say I'm completely satisfied with the code quality of the ROCMTarget yet, and it is clear that this was assembled from a few different sources. However, the sloppy error handling was masking real crash opportunities and has now been fixed.

### CODE REVIEW

File: compiler/plugins/target/ROCM/ROCMTargetUtils.cpp [hunk 1]
```diff
@@ -239,38 +239,42 @@ static LogicalResult linkModuleWithGlobal(llvm::Module *module,
...
+                                 StringRef targetChip, StringRef bitCodeDir,
+                                 unsigned linkerFlags,
+                                 llvm::TargetMachine &targetMachine) {
   // Early exit if Ukernel not supported on target chip.
-  if (!iree_compiler::hasUkernelSupportedRocmArch(targetChip))
-    return;
+  if (!iree_compiler::hasUkernelSupportedRocmArch(targetChip)) {
+    return mlir::emitError(loc)
+           << "ukernel " << enabledUkernelsStr
+           << " not supported on target chip: " << targetChip;
```
Comment (ScottTodd) [inline compiler/plugins/target/ROCM/ROCMTargetUtils.cpp]: could wrap this string in quotes too (helps with spotting empty strings, e.g. `ukernel  not supported` vs `ukernel '' not supported`)
```suggestion
    return mlir::emitError(loc)
           << "ukernel '" << enabledUkernelsStr
           << "' not supported on target chip: " << targetChip;
```

File: experimental/hip/tests/stablehlo_ops/CMakeLists.txt [hunk 2]
```diff
@@ -5,14 +5,8 @@
...
 unset(IREE_HIP_TEST_COMPILER_FLAGS)
-if(IREE_ROCM_LINK_BC)
-  list(APPEND IREE_HIP_TEST_COMPILER_FLAGS "--iree-rocm-link-bc=true")
-else()
-  list(APPEND IREE_HIP_TEST_COMPILER_FLAGS "--iree-rocm-link-bc=false")
-endif()
 list(APPEND IREE_HIP_TEST_COMPILER_FLAGS
-  "--iree-rocm-target-chip=${IREE_ROCM_TARGET_CHIP}"
-  "--iree-rocm-bc-dir=${IREE_ROCM_BC_DIR}")
+  "--iree-rocm-target-chip=${IREE_ROCM_TARGET_CHIP}")
```
Comment (ScottTodd) [inline experimental/hip/tests/stablehlo_ops/CMakeLists.txt:9]: fyi @sogartar (https://github.com/openxla/iree/pull/16769 also touches this code)

**Review by ScottTodd — APPROVED** [review]
Thank you!

File: compiler/plugins/target/ROCM/ROCMTarget.cpp [hunk 3]
```diff
@@ -51,13 +51,20 @@ namespace mlir::iree_compiler::IREE::HAL {
...
+  bool linkBitcode = true;
   int wavesPerEu = 0;
   std::string enableROCMUkernels = "none";
 
+  const std::string &getBitcodeDirectory() const {
+    if (bitcodeDirectory.empty()) {
+      return getDefaultBitcodeDirectory();
+    }
+    return bitcodeDirectory;
+  }
```
Comment (ScottTodd) [inline compiler/plugins/target/ROCM/ROCMTarget.cpp:66]: Also, /facepalm at how I had this written before... thanks for the function + private variable change here.

For LLVMCPUTarget, I put a helper class between the options and what the target actually sees: https://github.com/openxla/iree/blob/d05b4a16f826ca328f7e737ae473e2b581dc950f/compiler/plugins/target/LLVMCPU/LLVMTargetOptions.cpp#L656-L693

I guess that could be folded down to a single class by using a similar technique (private vars for all the internals, public functions to get values dependent on extra code)

Comment (stellaraccident) [reply compiler/plugins/target/ROCM/ROCMTarget.cpp:66]: Sometimes it takes a python programmer to properly encapsulate code.

**Review by benvanik — APPROVED** [review]
hell yeah

**Pull request merged**